### PR TITLE
beholder: a plugin for automatic keystroke, screenshot, and webcam captures

### PR DESCRIPTION
**Pull request by hdm**
This PR adds a new plugin (beholder) that enables regular capture of keystrokes, screenshots, and webcam snapshots from connected sessions. This plugin will also auto-migrate sessions into explorer.exe, which may not be the desired behavior, so care should be used.

The beholder plugin supports configuration parameters; these can be set via `beholder_conf option=val` before the plugin is started, or on the command line during startup with `beholder_start option=val`. The default configuration will automigrate to explorer.exe, start the keystroke logger, and start gathering screenshots every 30 seconds. The available configuration options and their default values are shown below:

```
      screenshot: true,
      webcam: false,
      keystrokes: true,
      automigrate: true,
      base: </.msf4/beholder/timestamp>
```

To use this plugin, load it and start it with the appropriate options:

```
msf> load beholder
msf> beholder_start base=/var/www/html/ freq=300
```

To stop beholder, use the `beholder_stop` command.

### CODE REVIEW

**Comment by hdm** [comment]
Once https://github.com/rapid7/metasploit-payloads/issues/102 is resolved, webcam capture can be re-enabled by default.

**Comment by wchen-r7** [comment]
That's pretty awesome. What do you think about having clipboard_monitor_start in here as well? Would that be appropriate for this plugin?

**Comment by hdm** [comment]
Clipboard and other stuff would be great; the only big issue with the current layout is that the first collector exception will cause further collectors to be skipped (sort of common in the real world when screenshot throws an error, webcam will be missed). For a big set of collectors, it probably needs to be reworked to wrap each of those in its own exception handler.

File: plugins/beholder.rb [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    def name
+      "Beholder"
+    end
+
+    def commands
+      {
+        'beholder_start'         => "Start capturing data",
+        'beholder_stop'          => "Stop capturing data",
+        'beholder_conf'          => "Configure capture parameters",
+      }
```
Comment (void-in) [inline plugins/beholder.rb]: Shouldn't we be using the ruby 1.9+ hash style here as well?

Comment (hdm) [inline plugins/beholder.rb]: Using the newer hash style for command dispatcher actually breaks things, for example:
```
    def commands
      {
        beholder_start: "Start capturing data",
        beholder_stop:  "Stop capturing data",
        "beholder_conf" =>  "Configure capture parameters",
      }
    end
```
```
help
[-] Error while running command help: comparison of Array with Array failed
```
```
msf exploit(handler) > beholder_start 
[-] Unknown command: beholder_start.
```

```
[*] Beholder Configuration
[*] ----------------------
[*]   screenshot: true
[*]   webcam: false
[*]   keystrokes: true
[*]   automigrate: true
[*]   base: /home/hdm/.msf4/beholder/2016-05-16.1463409298
[*]   freq: 30
[*]   idle: 0

```

Comment (wvu-r7) [inline plugins/beholder.rb]: @void-in: The hash keys need to be strings, not symbols.

**Comment by wchen-r7** [comment]
Works for me:

```
msf exploit(handler) > beholder_conf screenshot=true
[*] Beholder Configuration
[*] ----------------------
[*]   screenshot: true
[*]   webcam: false
[*]   keystrokes: true
[*]   automigrate: true
[*]   base: /Users/wchen/.msf4/beholder/2016-05-16.1463440161
[*]   freq: 30
[*]   idle: 0
msf exploit(handler) > beholder_start freq=10
[*] Beholder is logging to /Users/wchen/.msf4/beholder/2016-05-16.1463440161
msf exploit(handler) > ls /Users/wchen/.msf4/beholder/2016-05-16.1463440161
[*] exec: ls /Users/wchen/.msf4/beholder/2016-05-16.1463440161

2_WIN-6NH0Q8CJQVM_20160516-181049_screenshot.jpg
2_WIN-6NH0Q8CJQVM_20160516-181054_screenshot.jpg
2_WIN-6NH0Q8CJQVM_20160516-181105_screenshot.jpg
2_WIN-6NH0Q8CJQVM_20160516-181116_screenshot.jpg
session.log
msf exploit(handler) > 
```

File: plugins/beholder.rb (hunk 1 above)
Comment (void-in) [inline plugins/beholder.rb]: Thanks @hdm and @wvu-r7. And an excellent addition to the plugins especially for the long term campaigns.

Comment (wvu-r7) [inline plugins/beholder.rb]: Definitely!

**Comment by tdoan-r7** [comment]
# Release Notes

HP Data Protector 6.10/6.11/6.20 Install Service - This module allows you to exploit HP Data Protector, a backup and recovery system, to remotely upload files to the file share. Versions 6.10, 6.10, and 6.20 are vulnerable. Authentication is not required to exploit this vulnerability.

**Comment by Viss** [comment]
I found a fun .. I guess you'd call it a bug?
if you have multiple shells on the same box, and you have all of them automigrate to explorer or winlogon using beholder, the shells become super unstable and will die after 1-2 screenshots.
I'm not sure if there is a capability or mechanism within meterpreter or even msf to see "this box has a shell already, I don't need to use this new shell for beholder", but that sort of mechanism would reduce shell instability :D

**Comment by OJ** [comment]
> automigrate to explorer or winlogon using beholder

This looks like your problem to me! This kind of issue happens across the board when too many shells are in the same process. I don't think it's something specific to beholder, or even one thing that beholder does (eg. screenshots).

Screenshot functionality might not help though given that each call to get a screenshot a binary is uploaded.

> I'm not sure if there is a capability or mechanism within meterpreter or even msf to see "this box has a shell already, I don't need to use this new shell for beholder"

No there isn't. And there's room for mistakes here, as well as more forensic noise. On the Meterp side, if one Meterpreter can be found by another, it can be found by something else too. On the MSF side, it relies on the ability to always validate that a session exists on the same machine. This might be possible via machine ID and checking existing beholder sessions, but it sounds like a lot of effort for little gain.

The easy and sane (imho) solution here is not automatically firing up beholder on a new shell.

**Comment by Viss** [comment]
architecturally, beholder doesn't 'fire on specific shells'. it runs 'globally' on all shells. So it's basically either on or off. If it's running, then any shell that comes in becomes a shell that beholder runs against. 

You're right in that ultimately it's a 'too many shells migrated into the same process' issue - it's just weird that despite typing 'sessions' and being able to see what hosts have shells as the user of msf, meterp can't follow the same logic - e.g running 'sessions', seeing that there's more than one session for a given host, and just running against the first. I can do that manually as the user - it's weird that meterp or even msf can't do that as the tool itself.

**Comment by busterb** [comment]
so, the ask is that beholder fires once per specific host. well, we have enough data in the sessions to make sure that we only choose one session per host.

**Comment by Viss** [comment]
basically, yeah. The intention here is to not lose multiple shells at the cost of several screenshots and/or webcam pics.

**Comment by hdm** [comment]
@Viss that would make more sense as a new feature request versus commentary an older PR

**Comment by Viss** [comment]
fair point! I'll create a new feature request :D

**Pull request merged**